### PR TITLE
Shared albums should not consider the `require_link` property.

### LINKED ISSUE #1478: Shared albums which require a direct link for anonymous users are not visible to users with whom they are shared
### Detailed description of the problem

I realized during testing that not all shared albums I shared with a certain user would show up in the gallery once logging in with the other user. This was strange as it was only specific albums that would not show, always the same ones across users. I copied the URL to check if the album would be accessible that way, and sure it was: just would not show up in the gallery alongside the other albums.

At this point I realized the common thing between the albums were that they were all set to "Public". Somehow this seems to interfere with Albums::get, which does not list these albums in the API result.

### Steps to reproduce the issue

1. Create an album and share it with a user, confirm that it shows up in their gallery
2. Create another album, but set it to "Public" visibility (first three toggles on) before sharing it with an user
3. Confirm the album does not show up in that user's gallery, but is correctly accessible by its URL

### Browser and system

Lychee latest release built from Docker, running on LXC on Debian, up-to-date Firefox

### PR DESCRIPTION
**Pull request by ildyria**
Fixes #1478 

```php
/**
 *  - the user is the admin, or
 *  - the user is the owner, or
 *  - the album does not require a direct link and is shared with the user, or
 *  - the album does not require a direct link, is public and has no password set, or
 *  - the album does not require a direct link, is public and has been unlocked
 */
```

After further thoughts, this is pretty counter intuitive. We should have fixed that one a long time ago. :)

### CODE REVIEW

**Comment by ildyria** [comment]
I am not sure if I missed some other conditions.

File: app/Policies/AlbumQueryPolicy.php [hunk 1]
```diff
@@ -43,7 +43,7 @@ public function __construct(AlbumFactory $albumFactory, AlbumPolicy $albumPolicy
 	 *
 	 *  - the user is an admin
 	 *  - the user is the owner of the album
-	 *  - the album is shared with the user and the album does not require a direct link
+	 *  - the album is shared with the user
```
Comment (nagmat84) [inline app/Policies/AlbumQueryPolicy.php:46]: You have changed the comment of the method, but not the method itself, see lines 81-83 below. How so?

Maybe we should also have some tests which checks the new behavior.

**Review by nagmat84 — CHANGES_REQUESTED** [review]
You missed the conditions in line 82 and 360.

Moreover, I am in strong favor of tests which checks the new behavior. At the moment, we seem to have reached a point which is rather bug-free with respect to security issues (i.e. accidentally leaking a photo which should stay hidden). Any change in the SQL queries might break that and hence, tests would be wonderful.

Or we could just wait until https://github.com/LycheeOrg/Lychee/issues/1462 is ready. As I detailed out in https://github.com/LycheeOrg/Lychee/issues/1478 there might actually be other users who rely on exactly this feature, i.e. that albums can also be hidden from users with whom they are shared. (Although, I lack the imagination of a meaningful use-case for that.)

I am still wondering if we should make "hidden" a per-share option. @kamil4 raised the same question in https://github.com/LycheeOrg/Lychee/pull/1357#issuecomment-1215763003, 3rd bullet point. This being said, I am not sure, if we should change the behavior now or simply wait until https://github.com/LycheeOrg/Lychee/issues/1462. At least, the implementation works as designed and intended.

File: app/Policies/AlbumQueryPolicy.php (hunk 1 above)
Comment (ildyria) [reply app/Policies/AlbumQueryPolicy.php:46]: yeah I was thinking about adding some tests in that regard. At least to reproduce the bug and then kill it :)

**Comment by ildyria** [comment]
As per requested finished the PR changes and added tests. :)

**Comment by ildyria** [comment]
IDK what happened O.o but this is definitively not something I wanted to close. :'D

**Review by nagmat84 — APPROVED** [review]
LGTM, but untested. I only checked all the boolean conditions in the query (which was difficult enough :laughing:) and could not spot any remaining error.

@ildyria I assume you have tested it yourself and found no regression?

**Comment by ildyria** [comment]
> @ildyria I assume you have tested it yourself and found no regression?

Yes.

**Comment by ildyria** [comment]
Also the tests you added are making it pretty tight to create regressions :)

**Pull request merged**